libraries/AC_Avoidance/AP_Astar_RT.cpp

```diff
@@ -0,0 +1,686 @@
+        return RT_ASTAR_STATE_ERROR;
+    }
+
+    // time taken to find a path
+    const uint32_t path_finding_time = AP_HAL::millis() - visgraph_time;
+
+    if (_path_numpoints <= 2) {
+        // direct path exists
+        Write_RTAstar(RT_ASTAR_STATE_NOT_REQUIRED, _error_last_id, destination_new, nodes_time, visgraph_time, path_finding_time, (AP_HAL::millis()-start_ms));
+        return RT_ASTAR_STATE_NOT_REQUIRED;
+    }
+
+    Vector3f dest_pos;
+    uint16_t _path_idx_returned = 1;
+    if (get_shortest_path_point(_path_idx_returned, dest_pos)) {
+        // for the first point return origin as current_loc
+        Vector3f origin_pos;
+        if ((_path_idx_returned > 0) && get_shortest_path_point((_path_idx_returned-1), origin_pos)) {
+            // convert offset from ekf origin to Location
+            Location temp_loc(origin_pos, Location::AltFrame::ABOVE_ORIGIN);
+            origin_new = temp_loc;
+        } else {
+            // for first point use current loc as origin
+            origin_new = current_loc;
+        }
+
+        // convert offset from ekf origin to Location
+        Location temp_loc{Vector3f{dest_pos.x, dest_pos.y, 0.0f}, Location::AltFrame::ABOVE_ORIGIN};
+        destination_new = destination;
+        destination_new.lat = temp_loc.lat;
+        destination_new.lng = temp_loc.lng;
+
+        // Log
+        Write_RTAstar(RT_ASTAR_STATE_SUCCESS, _error_last_id, destination_new, nodes_time, visgraph_time, path_finding_time, (AP_HAL::millis()-start_ms));
+        return RT_ASTAR_STATE_SUCCESS;
+    }
+
+    // unreachable
+    return RT_ASTAR_STATE_ERROR;
+}
+
+// create nodes from OA DB obstacles. This will attempt to group closesly located obstacles.  Returns false on error or DB is empty
+bool AP_OART_AStar::create_oa_database_nodes()
+{
+// exit immediately if db is empty
+    AP_OADatabase *oaDb = AP::oadatabase();
+    if (oaDb == nullptr || !oaDb->healthy()) {
+        _error_last_id = AP_OA_RTAStar_Error::RT_ASTAR_ERROR_DB_ERROR;
+        return false;
+    }
+    // expand polygon point array if required
+    const uint16_t num_obstacles = oaDb->database_count();
+
+    if (num_obstacles == 0) {
+        // OA not needed
+        _error_last_id = AP_OA_RTAStar_Error::RT_ASTAR_ERROR_NONE;
+        return false;
+    }
+
+    // create a graph filed with close objects
+    Graph grouping_graph(num_obstacles);
+    group_oa_items(grouping_graph);
+
+    bool visited_node[num_obstacles] = {false};
+
+    // clear all points
+    _total_oa_db_numpoints = 0;
+
+
+    const uint16_t num_points_per_circle = OA_RTASTAR_NUM_POINTS_PER_OBJECT;
+
+    if (!oa_db_nodes.expand_to_hold(num_obstacles * num_points_per_circle)) {
+        _error_last_id = AP_OA_RTAStar_Error::RT_ASTAR_ERROR_OUT_OF_MEMORY;
+        return false;
+    }
+
+    // iterate through exclusion circles and create outer polygon points
```

Comment:
this comment needs updating perhaps because I think these are not polygon points.

Reply:
The obstacles can be treated as "exclusion circles" actually. I have updated the comment.